src/honey_badger/epoch_state.rs

```diff
@@ -205,15 +209,27 @@ where
             decryption: BTreeMap::default(),
             accepted_proposers: Default::default(),
             subset_handler: subset_handling_strategy.into(),
+            require_decryption,
             _phantom: PhantomData,
         })
     }
 
     /// If the instance hasn't terminated yet, inputs our encrypted contribution.
-    pub fn propose(&mut self, ciphertext: &Ciphertext) -> Result<Step<C, N>> {
-        let ser_ct =
-            bincode::serialize(ciphertext).map_err(|err| ErrorKind::ProposeBincode(*err))?;
-        let cs_step = self.subset.handle_input(ser_ct)?;
+    pub fn propose(
+        &mut self,
+        proposal: Vec<u8>,
+        rng: &mut Box<dyn Rng + Send + Sync>,
```

Comment:
I think Clippy is right here: why not make this a `&mut dyn Rng + Send + Sync`?

Reply:
If the function signature is `&mut (dyn Rng + Send + Sync)` then I need `.encrypt_with_rng(&mut Box::from(rng), ...)` later instead of just `.encrypt_with_rng(rng, ...)` because Rust complains when the argument to `encrypt_with_rng(...)` does not implement `Sized`.

I am a little fuzzy on the difference between the two. As I understand it, as it currently is, we have a reference to a pointer (I use 'pointer' here, and afterwards, loosely) on the stack which points to a heap allocated `Rng` trait object. Under the new signature, we would receive a reference to a stack allocated `Rng` trait object, then we create a heap allocated pointer TO that stack object, then we create a stack allocated reference to that heap allocated pointer and THAT is the argument we give to `encrypt_with_rng`.... 

From an efficiency standpoint, the former seems better (though I haven't gone through Godbolt or anything to see what it'd actually compile down to for release). Are there any ergonomics concerns that would offset that? Or, alternatively, is there another way to appease Rust that doesn't involve `Box`ing the reference argument?

Reply:
You can also omit the `Send + Sync` from the signature (sorry, I didn't think of that above): just `&mut Rng`, like all the other methods do that take an RNG argument, including the ones in the `rand` crate itself.

And you shouldn't need pass in a pointer to a pointer at all. `encrypt_with_rng(&mut *self.rng, …)` would borrow the RNG from its box, and only pass in a pointer to the RNG itself, I think?

Reply:
If I'm doing `&mut *self.rng` then I'm leaving the function signature of `propose` as `&mut Box<...>` instead of `&mut Rng` so we'd be ignoring clippy and this would all be moot anyway. (This also doesn't borrow the Rng from the box because we have `&mut Box<...>` so `&mut *rng` types back to `&mut Box<...>`.)

To reiterate, if I use `&mut Rng` in the function signature, it complains about `dyn rang::Rng` being unsized at the callsite of `encrypt_with_rng(...)`. Related, I'm not 100% clear how passing a `&mut Box<dyn Rng>` works in the first place since `encrypt_with_rng<R: Rng>(..., rng: R)` and AFAIK there isn't an `impl<T: Rng> Rng for Box<T>`. I guess somehow from the `Borrow` impl on `Box`  and then it autoderefs inside `encrypt_with_rng(...)`?

In any case, `encrypt_with_rng` requires something sized. AFAIK, that means we have a concrete type or a boxed trait object, right? So, unless there are ergonomic reasons that we'd prefer for `propose(...)` to accept an unboxed trait object, we might as well ignore clippy and require a boxed trait because we're gunna have to box it anyway.

Reply:
Right… but this works for me:

```rust
fn print_n<R: Rng>(rng: &mut R) {
    let n: u8 = rng.gen();
    println!("Random number: {}", n);
}

fn print_n_using_boxed(mut rng: Box<dyn Rng + Sync + Send>) {
    print_n(&mut rng);
}
```

So it looks like the function signature still doesn't need a `Box`, but you can just pass in `&mut self.rng`?

Reply:
This may all be moot because clippy is no longer reporting this as an issue in the recently released stable 1.30. So consider the rest of this more curiosity than necessity.

Where/how do you get that to work? https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=5026f41cd25e55fff07cc2fbd4c961de

1. I can't seem to get a direct translation to work in the playground using the current code as a base. It says that `Rng` is not implemented for `Box<dyn Rng>`. Which makes sense, I don't really understand why the current code compiles.
2. When I use `&mut dyn Rng` in the function signature, I get errors about it being unsized, which is what I've been saying this whole time. AFAIK, this is because `print_n` has an implicit `Sized` bound on `rng` which is why we need `Box`. And since it's defined in `threshold_encryption`, we can't change it inside this PR.

Reply:
Sorry, you're right, I mixed that up: let's just not make `propose`'s argument a fat pointer at all, just `&mut R` with a generic `R: Rng`. That's what all the other methods do, and it works with just `&mut self.rng`, I think.